### PR TITLE
[frameit] Ensure both texts use same baseline (issue #10440)

### PR DESCRIPTION
**Pull request by funnel20**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
Output: 
```
Finished in 6 minutes 46 seconds (files took 3.7 seconds to load)
4675 examples, 0 failures
```
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
Output: 
```
973 files inspected, no offenses detected
```
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.
Not applicable

### Motivation and Context

This solves issue https://github.com/fastlane/fastlane/issues/10440

The current implementation applies a `.trim` to both images for Keyword and Title.
Due to typography differences (for example the "L" that needs more vertical space than an "e") the generic baseline between 2 texts is lost. This can be seen in the excellent analysis of @janpio in this comment: https://github.com/fastlane/fastlane/issues/10440#issuecomment-348010988

_Before_
iphone6-01_remotetab_framed

This solution requests the calculated trim box and offset for each image **without applying it**.
By comparing the vertical top offsets and heights of both trim boxes, the vertical offset between them can be calculated. The box with the biggest offset will be compensated to equal the same of the other, so their text baseline will be equal. After this the height is made equal, so they share the same bottom offset.
Finally a `.crop` will be applied with the (compensated) trim boxes.

_After_
iphone6-01_remotetab_framed

Test results can be found in issue comment: https://github.com/fastlane/fastlane/issues/10440#issuecomment-348760923

Here the difference is clear, while highlighting the trim box in yellow:

Original code with `.trim`:
iphone6-01_remotetab_framed trim

My initial commit with `.crop` to align top:
iphone6-01_remotetab_framed vertical top

My robust solution with `.crop` for both align top and align bottom:
iphone6-01_remotetab_framed vertical bottom

It also works when only one of the images needs to be adjusted both on the top **and** bottom side. In this case for "en":
iphone6-01_remotetab_framed

### Description

The changes can be found in issue comment: https://github.com/fastlane/fastlane/issues/10440#issuecomment-348760717

Please note that:

- This will supersede PR #11096 as I committed that anonymously.
- I've no Ruby skills, so this might not be proper Production code. But the description and the code show the intended behaviour.
- Hence @janpio offered to help, so I added [WIP] to the title of the PR.

### CODE REVIEW

**Comment by janpio** [comment]
Remaining test failure is discussed here:
https://github.com/fastlane/fastlane/pull/11096#issuecomment-349593731
https://github.com/fastlane/fastlane/pull/11096#issuecomment-349602194

**Comment by janpio** [comment]
> The reason why I ended up with .identify was that I tried to do it with MiniMagick::Tool::Identify.new, but couldn't find a way to let that act on the image variable title_image instead of a file name.

Question is if `MiniMagick::Tool::Identify.new` would also use GraphicsMagick or ImageMagick... I don't really understand how it decides that :/ Only makes sense to switch if it does _not_ use GraphicsMagick which seems to be missing as a dependency.

Interesting enough `imagemagick` also doesn't seem to be defined as an dependency:
https://github.com/fastlane/fastlane/search?utf8=%E2%9C%93&q=imagemagick&type=
Does it maybe just work by accident (as it is maybe installed on macOS by default)?

**Comment by janpio** [comment]
I misinterpreted the error message:

` ImageMagick/GraphicsMagick is not installed`

means that the one of the both it tried to use is not installed.

**Comment by funnel20** [comment]
> Question is if MiniMagick::Tool::Identify.new would also use GraphicsMagick or ImageMagick... I don't really understand how it decides that :/ Only makes sense to switch if it does not use GraphicsMagick which seems to be missing as a dependency.

Line 327 (existing code before my edits) uses `title_image = MiniMagick::Image.open(empty_path)` to create the image. That did pass the tests before.

**Comment by funnel20** [comment]
> Interesting enough imagemagick also doesn't seem to be defined as an dependency:
> https://github.com/fastlane/fastlane/search?utf8=%E2%9C%93&q=imagemagick&type=
> Does it maybe just work by accident (as it is maybe installed on macOS by default)?

I have looked in my Terminal history. When I followed instructions from https://github.com/fastlane/fastlane/blob/master/YourFirstPR.md#checking-out-the-fastlane-repo
I needed to install Bundler and later in the local Fastlane repo run: `bundle update`
One of many output lines was:
`Using mini_magick 4.5.1`

**Comment by janpio** [comment]
Yes, but mini_magick is just a wrapper around ImageMagick and GraphicsMagick. 
But this is in the docs: https://docs.fastlane.tools/getting-started/ios/screenshots/#dependencies

Still strange why it didn't fail before...

I pulled @milch into this, he is looking into it.

**Comment by funnel20** [comment]
Maybe the test script is altered?

**Comment by milch** [comment]
@funnel20 I found the issue and fixed the tests directly. It should work now! We'll let it run for now to see if it does. 

It works locally because we all probably have imagemagick installed on our machines, which the CI doesn't have

**Comment by funnel20** [comment]
@milch Awesome, I see in your commit that the script just fakes some output. Thanks for your super fast assistance.

**Comment by funnel20** [comment]
I'm glad to see that the test passed.
@milch But someone here doesn't have a CLA...

**Comment by janpio** [comment]
That is something different and we can ignore - it just means more than 1 person commited and the maintainers have to check manually of it's okay to be merged. As @milch is the second person, it's all good :)

Can you remove the [WIP] from the title to indicate it is good to be reviewed?

**Review by milch — APPROVED** [review]
@funnel20 Great PR, thanks for your help 👍 

Sorry that you were having issues with the CLA in the previous PR. I definitely understand that the whole CLA business is frustrating for a first-time contributor. I hope now that everything is sorted out it won't discourage you from helping out on fastlane other Google-owned open source projects

To satisfy our @googlebot overlord, @funnel20 has agreed to have their code incorporated by submitting this PR and I, the only other commit author in this PR, obviously also agree to have my commit merged

**Comment by milch** [comment]
@funnel20 As @janpio said we can ignore the bot in this case. We both signed the CLAs and it is obvious that I agree to having the commit I made merged as well. 

The [WIP] was just because the tests didn't pass, right? The code looks good to me so if you confirm that we'll merge

**Comment by funnel20** [comment]
I have removed the WIP. It was because I have no Ruby skills and did the review together with @janpio in an earlier PR: https://github.com/fastlane/fastlane/pull/11096#pullrequestreview-80704510

**Review by janpio — APPROVED** [review]

**Comment by funnel20** [comment]
About the CLA. My blood pressure was finally a bit down. 
**But what did just happen?**
Now there is a failing CLA check and we can just override/ignore it by commenting that we all agree.
Or worse, @milch writes this on my behalf. Sorry?
This is unbelievable, unacceptable and probably illegal.

To satisfy the googlebot I would strongly recommend to make a valid, transparent and LEGAL option to let this check pass. Legal because I personally approved for God knows what when signing my CLA.
I expect that each submitter comments a similar command as `I signed it!`, for instance` I agree this PR` and the bot does the math if everyone has done so.

The whole idea of checks is that they should pass. During my endeavour of solving this issue, I hit 3 failing checks:

1. CLA
2. Rubocop error max. amount of Class lines
3. circle failing check.

I did really have to stretch my perseverance by counting till 10 a dozen of times.
But now the last check is just ignored. It's still not green. Or should I say RED.

**Comment by janpio** [comment]
Read what @googlebot writes in https://github.com/fastlane/fastlane/pull/11124#issuecomment-349645734

Background is that @googlebot is made for PRs with 1 committer, which covers 99% of PRs that come in. Only when someone with commit rights to the fastlane repo (as @milch and me for example) comes along and decide to commit to your PR, @googlebot bails out and posts this message.

It explicitly mentions that "It's up to you to confirm consent of the commit author(s) and merge this pull request when appropriate.". This is what we can now do to merge this PR: 

1. As @milch is a maintainer, I know he has a signed CLA. 
2. We know that you also have one, as @googlebot said so above.

So technically we could have merged your PR even before you had a valid CLA. But the message from @googlebot told us that this was not okay. Now he tells us, that we can use judgement and thinking to decide if we want to ignore the `x` the check provides right now.

No need to get riled up ;)

> The whole idea of checks is that they should pass.

Actually, the idea is that they check something. And depending on how they are configured, the can block a merge - or just inform about their state (and so optionally be ignored).

**Comment by funnel20** [comment]
So many checks, procedures and rules.
And the last one can stay red.
Seems a bad excuses to not have a proper procedure for this 1% use-case.

File: frameit/lib/frameit/editor.rb
Comment (ohayon) [inline frameit/lib/frameit/editor.rb]: Should we just remove these lines instead of leaving it commented?

File: frameit/lib/frameit/editor.rb
Comment (ohayon) [inline frameit/lib/frameit/editor.rb]: This seems a bit magical to know that we should index in to `3` and `1` below, is there any explanation we could add here to make this a bit more clear?

File: frameit/lib/frameit/editor.rb
Comment (ohayon) [inline frameit/lib/frameit/editor.rb]: it seems like we use index 1 and 3 quite a bit, could we stick them into their own, more descriptive variables?

**Review by ohayon — COMMENTED** [review]
This is a great change for sure. @funnel20 I just left a few comments that are definitely open for discussion. Nothing is too pressing and we totally understand if you just want to get this in quickly.

File: frameit/lib/frameit/editor.rb
Comment (funnel20) [reply frameit/lib/frameit/editor.rb]: I left it in for the code review as it was the root cause for the issue.
I agree that the commented `.trim` can be removed.
I suggest to merge the comment "Trimming the image will result..." in the comment a few lines below that starts with "Hence retrieve the calculated...", since it's the rational for the crop solution.

Comment (janpio) [reply frameit/lib/frameit/editor.rb]: Maybe make the comment "Naively trimming the image with `.trim` will..."?

Comment (ohayon) [reply frameit/lib/frameit/editor.rb]: That's cool @funnel20, you can make that whatever you'd like, just wanted to make sure you thought it was necessary.

File: frameit/lib/frameit/editor.rb
Comment (funnel20) [reply frameit/lib/frameit/editor.rb]: What about this extended comment:
```Ruby
# The .identify with format("%@") returns a string with "<width>x<height>+<offset_x>+<offset_y>". Extract these 4 parameters into an integer array, by using multiple string separators: the whitespace "\s", "x" and "+".
```

To also answer your next comment. Sure, we can define these indexes.
Do you have any naming convention for these kind of constants?
Example:
```Ruby
index_trim_box_width = 0
index_trim_box_height = 1
index_trim_box_offset_x = 2
index_trim_box_offset_y = 3
```

File: frameit/lib/frameit/editor.rb
Comment (funnel20) [reply frameit/lib/frameit/editor.rb]: See previous comment

Comment (funnel20) [reply frameit/lib/frameit/editor.rb]: Although..... Adding these 4 parameters would trigger the 320 lines Class limit again.
Or is there an overall Class with constants in the project?

File: frameit/lib/frameit/editor.rb
Comment (funnel20) [reply frameit/lib/frameit/editor.rb]: @janpio 👍

Comment (funnel20) [reply frameit/lib/frameit/editor.rb]: # Natively trimming the image with .trim will result in the loss of the common baseline between the text in all images. Hence retrieve the calculated trim bounding box without actually trimming. See: http://www.imagemagick.org/script/escape.php

Comment (ohayon) [reply frameit/lib/frameit/editor.rb]: thats great @funnel20 !

File: frameit/lib/frameit/editor.rb
Comment (ohayon) [reply frameit/lib/frameit/editor.rb]: I think code that is self explanatory is better than comments, I would say that you could do something like: 

```
trim_box_width = trim_values[0]
trim_box_height = trim_values[1]
trim_box_offset_x = trim_values[2]
trim_box_offset_y = trim_values[3]
```

File: frameit/lib/frameit/editor.rb
Comment (ohayon) [reply frameit/lib/frameit/editor.rb]: Lets not worry about the 320 lines for now. Lets make the code the best we can and we'll worry about trimming it down on our end.

File: frameit/lib/frameit/editor.rb
Comment (funnel20) [reply frameit/lib/frameit/editor.rb]: Understood.

File: frameit/lib/frameit/editor.rb
Comment (funnel20) [reply frameit/lib/frameit/editor.rb]: Do you mind if I restore the 1 line that you changed earlier to keep within the 320 limit? We both agreed that didn't result in the best code.

Comment (ohayon) [reply frameit/lib/frameit/editor.rb]: Thats cool! I plan to go in and make some changes before merging to keep the lines down. 😄

Comment (funnel20) [reply frameit/lib/frameit/editor.rb]: Can I just commit my changes simultaneously?

File: frameit/lib/frameit/editor.rb
Comment (funnel20) [reply frameit/lib/frameit/editor.rb]: Basically only need trim_box_height and trim_box_offset_y, this saves 2 lines already 👊

File: frameit/lib/frameit/editor.rb
Comment (ohayon) [reply frameit/lib/frameit/editor.rb]: Sure!

File: frameit/lib/frameit/editor.rb
Comment (funnel20) [reply frameit/lib/frameit/editor.rb]: Hmm, unfortunately I have to define them twice, because the trim box is used in 2 loops:
1. The loop where the image is created. To get the minimum top and maximum bottom offset of all images.
2. The loop where the trim boxes are compensated based on the earlier minimum top and maximum bottom offsets, to maintain the text baseline.

So now I have this with 2 times `# Get the height and vertical offset`:
```Ruby
        # The .identify with format("%@") returns a string with "<width>x<height>+<offset_x>+<offset_y>". Extract these 4 parameters into an integer array, by using multiple string separators: the whitespace "\s", "x" and "+":
        trim_values = calculated_trim_box.split(/[\sx+]/).map(&:to_i)
                            
        # Get the height and vertical offset:
        trim_box_height = trim_values[1]
        trim_box_offset_y = trim_values[3]

        # Get the minimum top offset of the trim box:
        if trim_box_offset_y < top_vertical_trim_offset
          top_vertical_trim_offset = trim_box_offset_y
        end

        # Get the maximum bottom offset of the trimbox, this is the top offset + height:
        if (trim_box_offset_y + trim_box_height) > bottom_vertical_trim_offset
          bottom_vertical_trim_offset = trim_box_offset_y + trim_box_height
        end

        # Store for the crop action:
        trim_boxes[key] = trim_values
      end

      # Crop images based on top_vertical_trim_offset to maintain text baseline:
      words.each do |key|
        # Get matching trim box:
        trim_box = trim_boxes[key]

        # Get the height and vertical offset:
        trim_box_height = trim_box[1]
        trim_box_offset_y = trim_box[3]
```

File: frameit/lib/frameit/editor.rb
Comment (janpio) [reply frameit/lib/frameit/editor.rb]: I actually really meant "Naively", as in "using the most obvious function that comes to mind to do the job of trimming". But yours probably works as well ;)

Comment (funnel20) [reply frameit/lib/frameit/editor.rb]: @janpio I see :-) 
It's a thinker though, as one might suspect a typo (like I did)

File: frameit/lib/frameit/editor.rb
Comment (ohayon) [reply frameit/lib/frameit/editor.rb]: Okay. So we have a couple options. I don't want you to keep having to make changes as you've already put in a great deal of work here. I am cool to merge it as is, and we can go back in and clean it up later.

The other option would be to create some sort of Ruby object that encapsulates these box dimensions and then have an initializer where it could be created from the array of `trim_boxes` that way it could be created and used a lot more efficiently *and* have the extra benefit of being abstracted into another class and therefor maybe help with the 320 line restriction.

What are your thoughts?

Comment (funnel20) [reply frameit/lib/frameit/editor.rb]: Option 2 is definitely preferred, as it improves on 3 subjects:
- readability
- 320 line check
- re-use

It would be great if you can make it, as I lack the Ruby skills. Or show me a similar file in the repo that I can use as example.

My functional thoughts:
1. The initializer uses the raw string from the .identify as input and retrieves it's 4 contents to create the trim box object with 4 populated properties: width, height, offset_x and offset_y
2. The current Class can read and change these object properties 
3. A method that returns the string format for the `.crop` method.

Comment (ohayon) [reply frameit/lib/frameit/editor.rb]: Yeah that sounds awesome and we are definitely happy to help with that!

**Comment by ohayon** [comment]
@funnel20 disregard that @googlebot message! Apologies for the noise with that.

**Comment by funnel20** [comment]
@ohayon No problem, I think it's my mistake as SourceTree committed the earlier commit of @milch again.

**Comment by funnel20** [comment]
The failing circle is expected, as I exceed the 320 lines (it's 325).
I want to fix that with the new trim box class, but first wanted to make this commit as the final solution in 1 class.

File: frameit/lib/frameit/editor.rb
Comment (funnel20) [reply frameit/lib/frameit/editor.rb]: For me just the outline of a class with an init method, a property and 1 method that returns a string is enough. Then I can populate it all myself.

What do you prefer as Class name?
Although currently everything is titled "trim box", maybe "crop box", "crop frame" or "crop canvas" are more appropriate?

**Comment by ohayon** [comment]
Sounds good. Let me know when you are in a place where you want us to take over and handle the refactor into that class! 🐰  🚀

**Comment by funnel20** [comment]
As mentioned in the (collapsed) review comment:

> If you can write (or point to) the outline of a class with an init method, a property and 1 method that returns a string is enough. Then I can populate it all myself.
> 
> What do you prefer as Class name?
> Although currently everything is titled "trim box", maybe "crop box", "crop frame" or "crop canvas" are more appropriate?

I'd like to start it myself, as this code generation gives me finally some satisfaction over all administrative hassle

**Comment by funnel20** [comment]
I guess this is a good candidate: https://github.com/fastlane/fastlane/blob/master/frameit/lib/frameit/screenshot.rb

**Comment by ohayon** [comment]
Ah got it, sorry I missed that message! 

Yeah, that screenshot class you linked to is a perfect candidate. I was just drafting up something like this:

```
class TrimBox
  attr_accessor :width
  attr_accessor :height
  attr_accessor :x_offset
  attr_accessor :y_offset

  def initialize(idenetify_string)
    @width = ...
    
  end

  def crop
    return "some string"
  end
end
```

**Comment by funnel20** [comment]
Awesome, try that tomorrow locally.
Two Ruby newbie questions:
1. In the screenshot class I sometimes see the @ before the variable, and sometimes not. What's the rule?
2. How to call the init from the Editor class?

**Comment by ohayon** [comment]
Those are two great questions: 

1. the `@` indicates that its an instance variable. Also worth noting that when you have an `attr_accessor` defined with the same name, any references to `variable_name` are really just getting out the value from `@variable_name`
2. just calling `new` on a class will call it's initializer (i.e. `TrimBox.new("some string")` calls through to `TrimBox#initialize`

🚀

**Comment by funnel20** [comment]
@ohayon I committed the new Trimbox class and the modified code in the Editor class.
Can you please verify it?

File: frameit/lib/frameit/trim_box.rb
Comment (funnel20) [inline frameit/lib/frameit/trim_box.rb]: I copied this logic from another Class.
When testing with faulty input, I see the Error in Terminal, but Frameit just continues.
It does not write a framed screenshot.

Is this intended behaviour?

File: frameit/lib/frameit/trim_box.rb
Comment (funnel20) [inline frameit/lib/frameit/trim_box.rb]: For readability I might guess the following is more clear:
`return "%dx%d+%d+%d" % [@width, @height, @offset_x, @offset_y]`

Do you agree?

Comment (ohayon) [reply frameit/lib/frameit/trim_box.rb]: I think what you have here looks great already! Using `"#{variable}` is definitely the preferred way to do string interpolation. 🚀

File: frameit/lib/frameit/trim_box.rb
Comment (ohayon) [reply frameit/lib/frameit/trim_box.rb]: It's possible this error is getting rescued somewhere and continuing execution. I would say leave it as is, this class looks 💯 !

**Review by ohayon — APPROVED** [review]

**Comment by ohayon** [comment]
@funnel20 - this is awesome. Thank you so much for putting in the time and effort and dealing with some of the stress around the process, I think I speak for our whole team that we have learned a lot and will work to make this easier and a better experience for contributors. We always strive to make contributing a pleasant experience for folks. 

We would love to see more from you. Please don't hesitate to reach out to us with more questions or for help implementing new features or fixing bugs. 🚀 🐙

**Comment by funnel20** [comment]
Excellent. 
Thanks to all for your patience with me and bringing me up to speed with Ruby.

**Pull request merged**
